File: frontend_app/middleware/withAdmin.ts
import { NextApiHandler, NextApiRequest, NextApiResponse } from 'next';
import { verifyJWT } from '@/lib/jwt';
import { prisma } from '@/app/helper/PrismaHelper';
import { AdminPayload } from '@/pages/api/user/admin';


export const withAdmin =
    (handler: NextApiHandler) =>
        async (req: NextApiRequest, res: NextApiResponse) => {
            const authHeader = req.headers.authorization;

            if (!authHeader || !authHeader.startsWith('Bearer ')) {
                return res.status(401).json({ error: 'Missing or invalid Authorization header' });
            }

            const token = authHeader.split(' ')[1];

            try {
                const payload = verifyJWT(token) as AdminPayload;
                const user = await prisma.admins.findFirst({
                    where: { user_id: payload.user_id },
                });
                if (!user) return res.status(401).json({ error: 'Invalid or expired token' });
                return handler(req, res);
            } catch (error) {
                return res.status(401).json({ error: 'Invalid or expired token' });
            }
        };

File: backend/middleware/withAdmin.py
from fastapi import Request, HTTPException
import jwt
from utils.constants.environment_keys import EnvironmentKeys
from utils.environment_manager import EnvironmentManager


async def verify_admin(request: Request):
    env_manager = EnvironmentManager()
    auth_header = request.headers.get("Authorization")
    
    if not auth_header or not auth_header.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="Missing or invalid Authorization header")
    
    token = auth_header.split(" ")[1]
    
    try:
        payload = jwt.decode(token, env_manager.get_key(EnvironmentKeys.SECRET_KEY.name), algorithms=["HS256"])
        return payload
    except jwt.PyJWTError:
        raise HTTPException(status_code=401, detail="Invalid or expired token") 